### ISSUE TITLE
Improving the search functionality & API Integration in MoreThanTV [MTV]

### ISSUE THREAD

**Issue by RoloSoze** [Jackett/Jackett#6783]
Searching MTV is different from searching most other well-known trackers. For example, if I wanted to search for a show like Sons of Anarchy, almost all other trackers will return the release name for each season. 

MoreThan.TV returns something like this: 'Sons.​of.​Anarchy.​S01.​1080p.​BluRay.​x264-​MTV' with MTV as opposed to the release group.

I get that this might have to do with the nested nature of Gazelle trackers, but the search functionality in other Gazelle trackers (e.g. Nebulance, Orpheus) works accurately and does return the correct release name.

Can this be fixed? It's a problem because when manually searching multiple trackers at once, the release I'm after may very well be on one of these two trackers but it's easy to miss if I don't look for it there specifically. 

Jackett version 0.12.1398.0
Mono version: N/A

**Comment by garfield69** [comment]
the awsome-hd indexer was written by the author in yaml knowing it was a piecemeal solution. #1481 #1943 #5225
its been agreed that the only solution is for the indexer to be rewritten in C#, but a volunteer was yet to contribute.

**Comment by RoloSoze** [comment]
Understood. I have sometime is able to do it in the near future. What about MTV?

**Comment by garfield69** [comment]
MTV is an indexer written in as a customised C# indexer nearly 3 years ago #477

Nebulance (transmitthenet) is even older, written 4 years ago as a customised C# indexer, probably one of the first indexer written for Jackett by the looks of the history tracks. 

Orpheus is also in C# but taps into the GazzelleTracker abstract, built about 2 years ago.

if you think MTV is a better fit for the GazzelleTracker, then a C# volunteer could use the Orpheus code as a model and convert MTV to use the GazzelleTracker abstract too.

**Comment by RoloSoze** [comment]
@garfield69 PTP is implemented perfectly and it's very similar to both AHD & MTV. Whoever worked on that is more than welcome to work on these two :)

**Comment by ngosang** [comment]
I will take a look if someone invites me. ngosang @ hotmail .es

**Comment by RoloSoze** [comment]
#7553 @ngosang Related :)

**Comment by ngosang** [comment]
> MoreThan.TV returns something like this: 'Sons.​of.​Anarchy.​S01.​1080p.​BluRay.​x264-​MTV' with MTV as opposed to the release group.

The release group is not available. You can make the search in the website and take a look. Maybe it's in the description page, but we can only get information from the torrent list.

If you try with other terms like `Siesta.Key.S03`, we are returning the group when it's available.

**Comment by RoloSoze** [comment]
> 
> 
> > MoreThan.TV returns something like this: 'Sons.​of.​Anarchy.​S01.​1080p.​BluRay.​x264-​MTV' with MTV as opposed to the release group.
> 
> The release group is not available. You can make the search in the website and take a look. Maybe it's in the description page, but we can only get information from the torrent list.
> 
> If you try with other terms like `Siesta.Key.S03`, we are returning the group when it's available.

The full release name, including the group name, is actually available, but like you said, it's on the torrent description page. I believe it was either @cadatoiva or @garfield69 who said that in C# trackers, like MTV and PTP, we can actually obtain information from the description page, but we should do it only if and when necessary, and I believe in this case, it's necessary, and it's actually being done in Jackett for PTP.

The example I mentioned above (Sons of Anarchy S01) appears like so in Jackett: 
![Sons MTV](https://user-images.githubusercontent.com/23747425/76712230-0aa3a300-66ed-11ea-9eef-54e7436b26c3.png)

Whoever implemented MTV in Jackett chose to append (-MTV) to the end of almost every season pack release name as a temporary solution until a C# developer comes along to do it right. 

Here's what the release name (folder name) looks like on MTV's website:

![Sons](https://user-images.githubusercontent.com/23747425/76712264-6e2dd080-66ed-11ea-859b-8af883822300.png)
 
The example you mentioned (Siesta Key S03) also has -MTV as the release name in Jackett. However, the correct release name shows up on the description page like so:
![Siesta](https://user-images.githubusercontent.com/23747425/76712295-bea52e00-66ed-11ea-9264-2ada6434bd60.png)

I'll give you another example. Sometimes Jackett won't even return -MTV. It'll just return something like 'Cheers Season 4' because that's how it's displayed on MTV's results page.

![Cheers MTV](https://user-images.githubusercontent.com/23747425/76712334-222f5b80-66ee-11ea-95e1-c0f86eba2b68.png)

The release name on the details page:
![Cheers](https://user-images.githubusercontent.com/23747425/76712337-307d7780-66ee-11ea-9bfb-0f6ecf0e37e4.png)

However, we still need to use the tags given (e.g. x264, SD) in case the full release name isn't provided. This is particularly true for SD release: 
![Cheers S01](https://user-images.githubusercontent.com/23747425/76712358-6de20500-66ee-11ea-9ad5-2aa2bd8678ce.png)

In this case, the tags would help to give more details about the release. It's not as big a problem for HD and 4K release, but I think the tags for those should still be used.

The best implementation example I've seen in Jackett for this kind of nested/grouped Gazelle trackers is the PTP implementation, so maybe @cadatoiva can help out with that. 

In the email I sent you, I explained how PTP results are returned in Jackett. PTP, MTV, AHD, Nebulance, Orpheus, all use the same Gazelle backend for their trackers.

Let me use the same example I emailed you for PTP to demonstrate the best way to return/display the search results in Jackett.

Here are the results for searching for Harry Potter and the Sorcerer's Stone (by IMDb ID) in Jackett. Focus on the 1st (largest) result:
![PTP 0](https://user-images.githubusercontent.com/23747425/76712692-7e47af00-66f1-11ea-9ad4-55d6261453d4.png)


Notice how the first part of the name is the full release name, and the second part [between square brackers] lists all the tags together separated by /

Now let's look at what these results look like on PTP's website:

Here's the search/browse page. That's where we got the tags from
![PTP1](https://user-images.githubusercontent.com/23747425/76712616-e77af280-66f0-11ea-8720-114c95979643.png)

And here's what the torrent details/description page looks like. That's where we got the full release name from:
![PTP2](https://user-images.githubusercontent.com/23747425/76712625-fe214980-66f0-11ea-997f-32466d903a0b.png)

Both MTV and AHD need to display the results in this exact same way for them to be informative, and it is possible since they both use the same structure as PTP.

Apply the same example for movie torrents on AHD and MTV as well. 

Please let me know if you have any questions.

**Comment by cadatoiva** [comment]
Unfortunately, I do not have access to these trackers, but @ngosang and @junglebus do. PTP is actually using a JSON api provided by the tracker that gives the full name from the query results. However, on the search results html page the user sees this information is still available inside attributes (something like `<div class="torrentName" long-name="full torrent name">Shortened name with...</div>`) So we still wouldn't have to follow a link to get the full name.

Maybe this can help with locating a similar solution on the AHD/MTV pages, being able to use `GetAttribute` to get the full name instead of relying on a TextContent that's incomplete.

**Comment by RoloSoze** [comment]
> 
> 
> Unfortunately, I do not have access to these trackers, but @ngosang and @junglebus do. PTP is actually using a JSON api provided by the tracker that gives the full name from the query results. However, on the search results html page the user sees this information is still available inside attributes (something like `<div class="torrentName" long-name="full torrent name">Shortened name with...</div>`) So we still wouldn't have to follow a link to get the full name.
> 
> Maybe this can help with locating a similar solution on the AHD/MTV pages, being able to use `GetAttribute` to get the full name instead of relying on a TextContent that's incomplete.

Since I'm not a coder myself, I'll let you, @cadatoiva,  @ngosang, and @garfield69 if he can help, discuss the details and possibilities, and maybe collaborate on this together. If APIs can solve this, @ngosang maybe you can have that discussion with the admins of AHD and MTV and see if they can offer something similar to what PTP offered.  The current search results from MTV are somewhat limited in Jackett and are extremely basic and limited in AHD so they definitely need to be worked on.

**Comment by cadatoiva** [comment]
MTV's source does seem to have some code in common with the abstract GazelleTracker class we have for the engine, so that part might be worth looking into. It would appear the PTP also has several similarities, so I'll have to look into that more myself. Gazelle abstract utilizes a JSON API, which gives more reliable results than HTML parsing.

**Comment by RoloSoze** [comment]
> 
> 
> MTV's source does seem to have some code in common with the abstract GazelleTracker class we have for the engine, so that part might be worth looking into. It would appear the PTP also has several similarities, so I'll have to look into that more myself. Gazelle abstract utilizes a JSON API, which gives more reliable results than HTML parsing. And AHD is currently in YML format, so I can't comment too much on that.

Out of the issues I'm personally facing with Jackett, this is probably the most pressing one as these are two top-tier trackers that aren't implemented right. @ngosang has access to both and is in the process of rewriting AHD in C#.

**Comment by cadatoiva** [comment]
Hopefully the Gazelle hint helps find the JSON API locations and makes implementing these two easier. I'll be saving this issue to remind me to look into abstracting up PTP into a Gazelle tracker and making sure that base class is working correctly too.

**Comment by RoloSoze** [comment]
> 
> 
> Hopefully the Gazelle hint helps find the JSON API locations and makes implementing these two easier. I'll be saving this issue to remind me to look into abstracting up PTP into a Gazelle tracker and making sure that base class is working correctly too.

Sounds good. Keep me posted, please.

**Comment by ngosang** [comment]
I didn't read all the messages, I will take a look when I have more time.

MTV and PTP looks similar but there are at least two big differences:
1. PTP has a parameter to return the results in JSON format that MTV doesn't. I'm sure that they include more details in the JSON than in the HTML page. https://passthepopcorn.me/torrents.php?searchstr=Harry.Potter&json=noredirect
2. In PTP the same URL allows to authenticate with ApiUser + ApiKey. In MTV we have to use user+password.

Although both are based on Gazelle, they either use different versions of Gazelle or have made a custom development.

In MTV we only have the HTML and we shouldn't get the details for each torrent. If one search return 100 results, we are going to perform 101 requests for each search in Jackett. This will cause problems for site administrators and users. It's better to talk with them so they add better API support.

**Comment by RoloSoze** [comment]
> 
> 
> I didn't read all the messages, I will take a look when I have more time.
> 
> MTV and PTP looks similar but there are at least two big differences:
> 
>     1. PTP has a parameter to return the results in JSON format that MTV doesn't. I'm sure that they include more details in the JSON than in the HTML page. https://passthepopcorn.me/torrents.php?searchstr=Harry.Potter&json=noredirect
> 
>     2. In PTP the same URL allows to authenticate with ApiUser + ApiKey. In MTV we have to use user+password.
> 
> 
> Although both are based on Gazelle, they either use different versions of Gazelle or have made a custom development.
> 
> In MTV we only have the HTML and we shouldn't get the details for each torrent. If one search return 100 results, we are going to perform 101 requests for each search in Jackett. This will cause problems for site administrators and users. It's better to talk with them so they add better API support.

Once you get the chance to read all the messages, I think you'll find that you and @cadatoiva are more or less on the same page about this issue. OK, if API support is the answer to this issue, let me have this discussion with the MTV admins first, and I'll loop in you in @ngosang. For AHD, we're just waiting for their reposne about APIs.

**Comment by ngosang** [comment]
@RoloSoze I think AHD is fixed in #7827 Could we rename this issue for MTV only? Or better, get the MTV important things and open a new issue.

**Comment by RoloSoze** [comment]
> 
> 
> @RoloSoze I think AHD is fixed in #7827 Could we rename this issue for MTV only? Or better, get the MTV important things and open a new issue.

Thank you so much. I will clean up this issue and make it AHD-specific. I'll test a few things when the update comes out. I'll move everything related to MTV to a new issue :)

**Comment by Taloth** [comment]
@garfield69 I was contacted by MTV staff that they're going to update their site and offered early access to be able to implement support before it goes live. Some stuff is still non-public so please contact me on discord (Jackett or Sonarr discords) and i'll put you in contact with them.
They're still inquiring internally about what api the new site supports, so not sure yet whether you need c# for it.

**Comment by RoloSoze** [comment]
@Taloth Thanks for the update! A few months ago, I was actually in touch with the MTV staff regarding this topic. My name, as well @ngosang & @cadatoiva are all on the list of names that are meant to get early access to the website, and also to have the discussion regarding API/tech to be used/issues, etc. @garfield69 was not on the list since he's not an MTV member, but he will be invited if MTV decides to go with something that @garfield69 can work with (i.e. not C#).

Having said that, I can take care of the coordination between us the Jackett team members & the MTV staff since I already started this conversation with them. Can you please shoot me an email regarding any details that you're aware of at this stage, including the MTV staff members that you're in touch with? You'll find my email address in my profile. Thanks.

**Comment by garfield69** [comment]
@RoloSoze  oh good, you beat me to a reply, thanks.

**Comment by Taloth** [comment]
yeah, I just needed to poke an active team member and discord doesn't appear to be active. I'll send an email with what I know.

**Comment by RoloSoze** [comment]
@Taloth Thank you.

@garfield69 I have been saving this invite for you anyway :) I'll invite you now regardless of whether you'll be able to contribute to the indexer.

**Comment by RoloSoze** [comment]
I got in touch with the MTV staff. Waiting for their instructions on how to proceed.

**Comment by garfield69** [comment]
>Update: Invite sent. 

no signs of invite email in any of my inboxes or junk/spam boxes.
garfieldsixtynine -at- gmail.com or garfield69@outlook.com

**Comment by RoloSoze** [comment]
@garfield69 Strange! I sent it to your Gmail. Just deleted it & sent a new one to your Outlook. Please check.

**Comment by garfield69** [comment]
no, nothing :-( perhaps the invitation system on the site is broken?

**Comment by RoloSoze** [comment]
> 
> 
> no, nothing :-( perhaps the invitation system on the site is broken?

Maybe. Reported to staff already.

**Comment by RoloSoze** [comment]
@garfield69 Yup, the invite system is not working at the moment. Will retry when it's fixed.

**Comment by RoloSoze** [comment]
The new URL for this tracker is:
https://www.morethantv.me/

@garfield69 I have sent an invite to your Gmail. Torrents have not been added to the new site but they will be soon. I'm not sure whether the invite system is up & running.

**Comment by RoloSoze** [comment]
@garfield69 Looks like the torrent page is up, at least partially. There is an announcement about that.

**Comment by garfield69** [comment]
some c# volunteer will need to rewrite the existing indexer to handle the new site torrent page.

**Comment by RoloSoze** [comment]
@garfield69 Are they not using Luminance? I thought Luminance implementation can be done with YAML?

**Comment by garfield69** [comment]
The engine may well be luminance, but clearly the engine supports customising pages.
The torrent page at the site has a season title that comprises multiple download links.
yaml cannot process this. requires C# support.

**Comment by RoloSoze** [comment]
@cadatoiva @ngosang Can either of you work on rewriting this indexer now that it's up & running? Thank you!

**Comment by ilike2burnthing** [comment]
@cadatoiva hasn't been around since July.

**Comment by RoloSoze** [comment]
> 
> 
> cadatoiva hasn't been around since July.

Yup, I'm kinda hoping for a triumphant return 😃

**Comment by maddestructor** [comment]
I can probably try to give it a go in my free time if you guys want. Will take some time as I'm pretty busy at work at the moment and I don't know jackshit about jackett's code (see what I did there 😛) so will have to discover it as well.

**Comment by RoloSoze** [comment]
> 
> 
> I can probably try to give it a go in my free time if you guys want. Will take some time as I'm pretty busy at work at the moment and I don't know jackshit about jackett's code (see what I did there 😛) so will have to discover it as well. I'm the one who opened the TLS issue by the way.

@maddestructor Yup, go ahead, please! Thanks!

**Comment by maddestructor** [comment]
Just wanted to let you guys know that I'm all set up to start and that it took me 15 minutes to get Jackett's code pulled, built and running so it shouldn't be too bad. I am currently waiting on the staff of MTV to send me any details that could help me regarding their documentation or the underlying technologies they're using for the new platform.

Once I receive an answer (which shouldn't take too long), I'll start writing a C# indexer (if needed). I've looked at the code a bit and it seems fairly easy so it shouldn't take too much time. Also, good job to the maintainers of the code, it's pretty neat and clean compared to some of the stuff I often see elsewhere ;)

**Comment by ryanm352** [comment]
@maddestructor -- challenge accepted. Still needs more testing.
https://github.com/ryanm352/Jackett/commit/ce73718380bc9c6403df0f3f4d1dc37f2c65316f

**Comment by garfield69** [comment]
v0.17.699

**Comment by garfield69** [comment]
v0.17.737
